src/drivers/kinetis/i2c.c

```diff
@@ -0,0 +1,310 @@
+
+                    /* address write post repeated start */
+                    ch->txrx = I2C_WRITING;
+                    ch->seq++;
+                    base->D = (*ch->seq).addr;
+                }
+                /* no repeated start */
+                else {
+                    goto i2c_isr_stop;
+                }
+                break;
+            case 1:
+                /* do not ack the final read */
+                base->C1 |= I2C_C1_TXAK_MASK;
+                *ch->received_data++ = base->D;
+                break;
+            default:
+                /* simply perform a read */
+                *ch->received_data++ = base->D;
+                break;
+        }
+        /* decrement the number of reads */
+        ch->reads_ahead--;
+    }
+    /* handle writing */
+    else {
+        /* check if sequence end */
+        if(ch->seq == ch->seq_end) {
+            goto i2c_isr_stop;
+        }
+
+        /* check if NACK */
+        if(base->S & I2C_S_RXAK_MASK) {
+            goto i2c_isr_error;
+        }
+
+        /* check if we have a restart */
+        if((*ch->seq).opcode == I2C_RESTART) {
+            /* generate repeated start and ensure TX */
+            base->C1 |= (I2C_C1_RSTA_MASK | I2C_C1_TX_MASK);
+
+            /* address write post repeated start */
+            ch->seq++;
+            base->D = (*ch->seq).addr;
+        }
+        /* not a repeated start */
+        else {
+            /* handle a read */
+            if((*ch->seq).opcode == I2C_READ) {
+                /* set channel status to read */
+                ch->txrx = I2C_READING;
+
+                /* at least one read ahead */
+                ch->reads_ahead = 1;
+
+                /* parse the number of reads */
+                while(((ch->seq + ch->reads_ahead) < ch->seq_end) &&
+                        ((*(ch->seq + ch->reads_ahead)).opcode == I2C_READ)) {
+                    ch->reads_ahead++;
+                }
+
+                /* switch to rx */
+                base->C1 &= ~I2C_C1_TX_MASK;
+
+                if(ch->reads_ahead == 1) {
+                    /* do not ACK the final read */
+                    base->C1 |= I2C_C1_TXAK_MASK;
+                }
+                else {
+                    /* ACK every other read */
+                    base->C1 &= ~(I2C_C1_TXAK_MASK);
+                }
+
+                /* dummy read without incrementing received data pointer */
+                *ch->received_data = base->D;
+                ch->reads_ahead--;
+            }
+            /* handle a write */
+            else {
+                base->D = (*ch->seq).value;
+            }
+        }
+    }
+    ch->seq++;
+    goto i2c_isr_exit;
+
+i2c_isr_stop:
+    /* generate a STOP, switch to RX, disable interrupts */
+    base->C1 &= ~(I2C_C1_MST_MASK | I2C_C1_TXAK_MASK | I2C_C1_IICIE_MASK);
+
+    /* call the callback function on success if it exists */
+    if(ch->callback) {
+        (*ch->callback)(ch->args);
```

Comment:
This is dangerous.  Like we learned in the uart code, calling functions in an ISR is lethal if the ISR does not block.  Hopefully that is not the case here, because this function should be fast, but if someone decided to say, send a log message in this callback, that would be terrible.
Possible solutions:
  - schedule the callback to be called later (requires process switcher)
  - `cpsid` (callback) `cpsie` iff the callback exists
  - I2C interrupts are disable-able without destroying state here (I think), we might be able to ensure the i2c module cannot preempt itself.

Reply:
I think we can ignore it for now. As of right now, nothing is using the callback functionality. Ideally, we would have the process switcher to handle this problem. Additionally, the ISR `cpsid`, `cpsie` would prevent any problems here.

Reply:
Yeah, I agree this can be ignored for the time being.  Worth noting for later, though.